Update UpCloud config entry data on successful reconfig

scop opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Handle UpCloud reconfig by checking unique id, update on match.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

scop committed:
Update UpCloud config entry data on successful reconfig

emontnemery reviewed (commented):
The change looks good, but please add a test.
Also, does the PR description match the implementation, the implementation aborts the flow if there's a unique id collision, the PR description talks about "update on match" during reconfig?

scop committed:
Add already configured test

scop committed:
Improve success case request mocking

So that the data fetch that might happen in the background after success
won't trash the test log with misleading errors.

scop commented:
Test added, description refined.

frenck commented:
Ideally, this should implement a re-auth flow.

scop commented:
Maybe not, however there's nothing that would actually prevent users from reconfiguring it like this, and I think what's implemented here is the sane behavior for such cases.

scop commented:
This change updates data only on successful auth, i.e. when reconfiguring with the _correct_ password. It will not change anything about trying to configure with an incorrect password. Doing so will keep failing just like it used to before this change.

Without this, when the user reconfigures using an username for which we already have a service for, the end result is the same as with this change, the data for an existing service with the same username is updated. The difference is that without this the user is told "Created configuration for ...", with every indication that a new service was created, which is confusing and untrue. With this change, a notification that the service already existed is displayed instead (and the data updated), which is accurate.

frenck commented:
> This change updates data only on successful auth, i.e. when reconfiguring with the correct password. It will not change anything about trying to configure with an incorrect password. Doing so will keep failing just like it used to before this change.

Understood.

> Without this, when the user reconfigures using an username for which we already have a service for and correct credentials for it, the end result is the same as with this change, the data for an existing service with the same username is updated. 

Understood, but we have existing logic for that, which is re-auth handling.

>  With this change, a notification that the service already existed is displayed instead (and the data updated), which is accurate.

I'm fine with that co-existing next to the regular re-auth handling.

MartinHjelmare reviewed (approved):
Looks good! It would be good to also add a reauth flow in another PR, but that doesn't stop this PR, I think.

scop committed:
Simplify already configured test

Thanks-to: Martin Hjelmare <marhje52@gmail.com>

scop commented on tests/components/upcloud/test_config_flow.py:
```diff
@@ -105,3 +106,31 @@ async def test_options(hass: HomeAssistant) -> None:
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+    stored_entry = hass.config_entries.async_get_entry(config_entry.entry_id)
```
Thanks!

frenck commented:
> Looks good! It would be good to also add a reauth flow in another PR, but that doesn't stop this PR, I think.

Sorry, I don't agree with that.

See also: https://github.com/home-assistant/core/pull/68718#issuecomment-1091720201

MartinHjelmare commented:
If we allow this feature, we should allow to add it separately from the reauth flow as it's two different features and we want small contained PRs.

Regarding allowing it or not, the motivation I see is it allows the user to change password without waiting for an auth failure which for polling integrations can take time. If we don't think that's a good enough reason I don't see much use for it and we can say that reauth flow is the only allowed way to change password.
